pkg/controller/awsfederatedaccountaccess/awsfederatedaccountaccess_controller.go

```diff
@@ -0,0 +1,368 @@
+func (r *ReconcileAWSFederatedAccountAccess) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling AWSFederatedAccountAccess")
+
+	// Fetch the AWSFederatedAccountAccess instance
+	currentFAA := &awsv1alpha1.AWSFederatedAccountAccess{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, currentFAA)
+	if err != nil {
+		if k8serr.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	// If the state is applied don't do anything
+	if currentFAA.Status.State == awsv1alpha1.AWSFederatedAccountStateReady || currentFAA.Status.State == awsv1alpha1.AWSFederatedAccountStateFailed {
+		return reconcile.Result{}, nil
+	}
+
+	// Get a list of all available roles
+	federatedRoleList := &awsv1alpha1.AWSFederatedRoleList{}
+	if r.client.List(context.TODO(), &client.ListOptions{}, federatedRoleList); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Create map to make finding role easier
+	roleMap := make(map[string]awsv1alpha1.AWSFederatedRole)
+	for _, role := range federatedRoleList.Items {
+		roleMap[role.Name] = role
+	}
+
+	// See if requested role exists
+	if _, ok := roleMap[currentFAA.Spec.AWSFederatedRoleName]; !ok {
+		SetStatuswithCondition(currentFAA, "Requested role does not exist", awsv1alpha1.AWSFederatedAccountFailed, awsv1alpha1.AWSFederatedAccountStateFailed)
+		reqLogger.Error(ErrFederatedAccessRoleNotFound, fmt.Sprintf("Resquested role '%s' not found", currentFAA.Spec.AWSFederatedRoleName))
+
+		err := r.client.Status().Update(context.TODO(), currentFAA)
+		if err != nil {
+			reqLogger.Error(err, fmt.Sprintf("Status update for %s failed", currentFAA.Name))
+			return reconcile.Result{}, err
+		}
+
+		return reconcile.Result{}, nil
+
+	}
+
+	// Get aws client
+	awsClient, err := awsclient.GetAWSClient(r.client, awsclient.NewAwsClientInput{
+		SecretName: currentFAA.Spec.AwsCredentialSecret.Name,
+		NameSpace:  currentFAA.Spec.AwsCredentialSecret.Namespace,
+		AwsRegion:  "us-east-1",
+	})
+	if err != nil {
+		awsClientErr := fmt.Sprintf("Unable to create aws client for region ")
+		reqLogger.Error(err, awsClientErr)
+		return reconcile.Result{}, err
+	}
+
+	// here create the custom policy in the cluster account
+	err = r.createIAMPolicy(awsClient, roleMap[currentFAA.Spec.AWSFederatedRoleName], *currentFAA)
+	if err != nil {
+		// if we were unable to create the policy fail this CR.
+		SetStatuswithCondition(currentFAA, "Failed to create custom policy", awsv1alpha1.AWSFederatedAccountFailed, awsv1alpha1.AWSFederatedAccountStateFailed)
```

Comment:
Should this be set to failed? What if there's an API issue and this needs to just be retried for it work? As of now, this would permanently make this CR dead, i.e. never reconciled again without human intervention due to line 111.

Reply:
We can create a retry in the future , this is just the MVP. We also before this goes into production have cleaning steps whenever this is failed or else it will not work on next reconcile if the objects attempted to be created have the same name.